### DISCUSSION TITLE
Next.js 13 - Magic Link Issue

### DISCUSSION THREAD

**Discussion by margsenjo** [supabase/supabase#15526] [Questions] [answered]
Hey guys,

I am relatively new to Supabase and Next.js and I am encountering an issue that I cant come across. I am trying to set up the login by magic link and everything until sending the email works fine. After clicking the magic link though, the user is not logged in. Here's some context:

**_/auth/callback.tsx:_**
```typescript
import { createRouteHandlerClient } from '@supabase/auth-helpers-nextjs'
import { cookies } from 'next/headers'
import { NextResponse } from 'next/server'

export async function GET(request: Request) {
  // The `/auth/callback` route is required for the server-side auth flow implemented
  // by the Auth Helpers package. It exchanges an auth code for the user's session.
  // https://supabase.com/docs/guides/auth/auth-helpers/nextjs#managing-sign-in-with-code-exchange
  const requestUrl = new URL(request.url)
  const code = requestUrl.searchParams.get('code')

  if (code) {
    const supabase = createRouteHandlerClient({ cookies })
    await supabase.auth.exchangeCodeForSession(code)
  }

  // URL to redirect to after sign in process completes
  return NextResponse.redirect(requestUrl.origin)
}

``` 

**_Login Page:_**
```typescript
'use client'

import { useState } from 'react'
import { useRouter } from 'next/navigation'
import { createClientComponentClient } from '@supabase/auth-helpers-nextjs'
import Link from 'next/link'
import { toast } from 'react-toastify'

export default function Login() {
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [resetEmail, setResetEmail] = useState('')
  const router = useRouter()
  const supabase = createClientComponentClient()
  const [errorMessage, setErrorMessage] = useState('')
  const [isLoading, setIsLoading] = useState(false);



  const handleSignIn = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    setIsLoading(true);
    const { data, error } = await supabase.auth.signInWithOtp({
      email,
      options: {
        emailRedirectTo: `${new URL(location.href).origin}/auth/callback`
      },
    })

    if (error) {
      toast(error.message)
    } else {
      setErrorMessage('') // clear any existing error messages
      router.push('/')
      router.refresh()
    }
    setIsLoading(false);
  }  

  return (
    <div className="flex-1 flex flex-col w-full px-8 max-w-md justify-center gap-2">
      <div className='p-8 rounded-2xl bg-apoio-secondary border-[#2c2c2c] border'>
        <div className='login_header w-full flex flex-col justify-center items-center'>
        <img className="p-3 max-w-[96px]" src="/img/apoio-main-logo-white.svg" />
        <h1 className='my-5'>Client Portal</h1>
        </div>
        <div className='mb-16 border-b border-[#2c2c2c] w-60'></div>
  <form
    className="flex-1 flex flex-col w-full justify-center gap-2 text-foreground"
    onSubmit={handleSignIn}
  >
    <label className="text-md" htmlFor="email">
      E-Mail
    </label>
    <input
      required
      type='email'
      className="rounded-md px-4 py-2 bg-inherit border mb-6 focus:ring-[#00fc6d] focus-visible:outline-0 focus-within:ring-2"
      name="email"
      onChange={(e) => setEmail(e.target.value)}
      value={email}
      placeholder="you@example.com"
    />
   <button  type='submit' className="flex align-center justify-center bg-apoio-green rounded-full hover:bg-transparent transition hover:text-[#fefefe] border-2 border-[#00fc6d] px-4 py-2 apoio-main mb-6" disabled={isLoading}>
          {isLoading ? (
            <svg width="20" height="20" viewBox="0 0 38 38" xmlns="http://www.w3.org/2000/svg" stroke="currentColor">
              <g fill="none" fillRule="evenodd">
                <g transform="translate(1 1)" strokeWidth="2">
                  <circle strokeOpacity=".5" cx="18" cy="18" r="18"/>
                  <path d="M36 18c0-9.94-8.06-18-18-18">
                    <animateTransform
                      attributeName="transform"
                      type="rotate"
                      from="0 18 18"
                      to="360 18 18"
                      dur="1s"
                      repeatCount="indefinite"/>
                  </path>
                </g>
              </g>
            </svg>
          ) :
      'Log-In Link bekommen' }
    </button>
  </form>
      </div>
    </div>
  )
}

``` 

**Middleware:**
```typescript
import { createMiddlewareClient } from '@supabase/auth-helpers-nextjs'
import { NextResponse } from 'next/server'

import type { NextRequest } from 'next/server'

export async function middleware(req: NextRequest) {
  const res = NextResponse.next()
  const supabase = createMiddlewareClient({ req, res })
  const PUBLIC_FILE = /\.(.*)$/;
  if (
    req.nextUrl.pathname.startsWith("/_next") || // exclude Next.js internals
    req.nextUrl.pathname.startsWith("/api") || //  exclude all API routes
    req.nextUrl.pathname.startsWith("/static") || // exclude static files
    req.nextUrl.pathname.startsWith("/password") || // exclude static files
    req.nextUrl.pathname.startsWith("/login") || // exclude static files

    PUBLIC_FILE.test(req.nextUrl.pathname)  // exclude all files in the public folder

  )
    return NextResponse.next();

  const {
    data: { user },
  } = await supabase.auth.getUser()

  // if user is signed in and the current path is / redirect the user to /account
  if (user && req.nextUrl.pathname === '/') {
    return NextResponse.redirect(new URL('/account', req.url))
  }

  // if user is not signed in and the current path is not / redirect the user to /
  if (!user && req.nextUrl.pathname != '/login') {
    return NextResponse.redirect(new URL('/login', req.url))
  }

  return res
}

``` 

For some reason, the user is not logged in after clicking the link. Using a regular login with email and password however, worked fine for me.

Any tips/guidances are highly appreciated!

**Comment by gokulkrishh** [comment]
@margsenjo Previously in next 13 or before with pages, supabase doesnt need any code to login with magic link it was doing it with authProvider in reactJS i guess. Now with appDir, we need to setSession using 

```js
 const { data, error } = supabase.auth.setSession({
    access_token,
    refresh_token
  })
```
[Doc link](https://supabase.com/docs/reference/javascript/auth-setsession). 

Correct me if i am wrong. I am using magic link in [expense.fyi](https://github.com/gokulkrishh/expense.fyi) site. It works fine with pages but with appDir it doesnt work automatically (my guess is now its using cookies so we need to set it). I needed to manually set it.

And also exchangeCodeForSession in /auth/callback.tsx is for regular login with password. For magic link you need handle it.

**Comment by taivo** [comment]
Hi @margsenjo were you able to find a solution?

**Comment by porkopek** [comment, marked as answer]
This is the new way to do everything related to emails (reset, magic link, etc.)

https://supabase.com/docs/guides/auth/server-side/email-based-auth-with-pkce-flow-for-ssr

**Reply by sujjeee** [reply]
Hey @porkopek 

can you help me with this ??

I'm using Next.js 14 with the app router.

This is how I am generating a magic link:

```
const { data, error } = await supabase.auth.admin.generateLink({
  type: "magiclink",
  email: email,
  options: {
    redirectTo: `${env.NEXT_PUBLIC_APP_URL}/auth/confirm`,
  },
});
```

 And this is my auth confirmation code, located at "auth/confirm/route.ts" (copied from the Supabase documentation). 
auth/confirm/route.ts 

 This is what the magic link looks like (when we console .log data): 
 
`'https://arsbplllhwdwohaaysmv.supabase.co/auth/v1/verify?token=9efed3a9e185a2d537a30344ff1c93ea0d8e2adaca256ee8738d842f&type=magiclink&redirect_to=http://localhost:3000/auth/confirm', `

 But the issue is when I click on the link, it redirects me to this. Why is that happening? 
 
`http://localhost:3000/auth/auth-code-error#access_token=eyJhbGciOiJIUzI1N...&expires_at=1710407536&expires_in=3600&refresh_token=KW0MwfdThrv5yUpOf04ulw&token_type=bearer&type=magiclink `

 Please Help if you know how to fix this.  🙏

**Comment by gokulkrishh** [comment]
With Safari 17.2, Login cookies are shared by the web application added to home screen. We can now add the app to home screen and use the same cookies to stay logged in.

> @margsenjo Previously in next 13 or before with pages, supabase doesnt need any code to login with magic link it was doing it with authProvider in reactJS i guess. Now with appDir, we need to setSession using
> 
> ```js
>  const { data, error } = supabase.auth.setSession({
>     access_token,
>     refresh_token
>   })
> ```
> 
> [Doc link](https://supabase.com/docs/reference/javascript/auth-setsession).
> 
> Correct me if i am wrong. I am using magic link in [expense.fyi](https://github.com/gokulkrishh/expense.fyi) site. It works fine with pages but with appDir it doesnt work automatically (my guess is now its using cookies so we need to set it). I needed to manually set it.
> 
> And also exchangeCodeForSession in /auth/callback.tsx is for regular login with password. For magic link you need handle it.
